### ISSUE TITLE
[BUG]: trying to catch a py::type_error makes the compilation crash with Clang on Windows

### ISSUE THREAD

**Issue by MatthieuHernandez** [pybind/pybind11#3746]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

The compiler Clang v13.0.0 craches at the linking step on Windows when add to following line to my code : 
```cpp
 try {
    auto casted_values = py::cast<std::vector<int32_t>>(values); // Here values is a py::list&.
  } catch (const py::cast_error& e) {
    throw py::type_error("this method expects integers as param");
  }
```

The error is : 
```
LLVM ERROR: Associative COMDAT symbol '??0builtin_exception@pybind11@@QEAA@PEBD@Z' does not exist.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
```

The strange thing is that if I recompile the same code but keep the CMake cache, it works. 🤔 
It also works perfectly under Linux with CGG.

Does anyone have an explanation for this?

EDIT : It seems that the problem comes from the try catch but I have no idea why.

**Comment by MatthieuHernandez** [comment]
Yes but this bug appends only when I try to catch a pybind11 error, if I throw an std::runtime_error it works perfectly.

**Comment by MatthieuHernandez** [comment]
I just updated Pybind11 to v2.9.1 and that solved the problem. Sorry for the inconvenience :sweat_smile:

**Comment by henryiii** [comment]
A change to pybind11 isn't going to change anything, this is still a clang on Windows bug, I'd recommend opening it with them and linking here. A compiler should never segfault when compiling.

>  catch a pybind11 error, if I throw a `std::runtime_error` instead it works perfectly.

That just means their bug has to do with something custom in our error. I expect they test on `std::runtime_error`. Still their bug. If the compiler stops crashing and instead complains about something, then we can work it out.
